source/server/connection_handler_impl.cc

```diff
@@ -95,6 +95,29 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
           details->typed_listener_)) {
     tcp_listener_map_by_address_.insert_or_assign(
         config.listenSocketFactory().localAddress()->asStringView(), details);
+
+    auto& address = details->address_;
+    // If the address is IPv6 and isn't v6only, parse out the ipv4 compatible address from the ipv6
+    // address and put an item to the map. Then this allows the `getBalancedHandlerByAddress`
+    // can match the ipv4 any-address also.
+    if (address->type() == Network::Address::Type::Ip &&
+        address->ip()->version() == Network::Address::IpVersion::v6 &&
+        !address->ip()->ipv6()->v6only()) {
+      if (address->ip()->isAnyAddress()) {
+        tcp_listener_map_by_address_.insert_or_assign(
+            Network::Address::Ipv4Instance(address->ip()->port()).asStringView(), details);
+      } else {
+        auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
+        // Remove this check when runtime flag
```

Comment:
It may be better to leave the check in, and trigger an ENVOY_BUG when it fails, after the runtime check is removed. That seems better than crashing, and this check is inexpensive.

Reply:
When `envoy.reloadable_features.strict_check_on_ipv4_compat` is false, should we go back to the original behavior instead of let Envoy stop?

Reply:
ENVOY_BUG won't stop in a production build, so I think it might be worth it to do it in this case.

Reply:
I see, that makes sense, thanks!